cpp/include/raft/spatial/knn/detail/ann_ivf_flat_kernel.cuh

```diff
@@ -0,0 +1,1227 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// #define USE_FAISS
+
+#include "../ann_common.h"
+#include "ann_utils.cuh"
+#include "topk/warpsort_topk.cuh"
+
+#include <raft/common/device_loads_stores.cuh>
+#include <raft/core/logger.hpp>
+#include <raft/cuda_utils.cuh>
+#include <raft/cudart_utils.h>
+#include <raft/distance/distance.cuh>
+#include <raft/pow2_utils.cuh>
+
+#ifdef USE_FAISS
+#include <faiss/gpu/utils/Comparators.cuh>
+#include <faiss/gpu/utils/Select.cuh>
+#endif
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <optional>
+#include <raft/core/mdarray.hpp>
+
+namespace raft::spatial::knn::detail {
+
+constexpr int kThreadsPerBlock = 128;
+
+namespace md = std::experimental;
+
+template <typename T>
+struct ivf_flat_index {
+  using row_major = md::layout_right;
```

Comment:
> Would it be ok if we exposed layout_rightor row_major in the raft namespace as well?

Yes, absolutely I think having both exposed is fine. As the `mdspan`/`mdarray` is a somewhat natural analogue to Numpy's `ndarray`, we preferred to also expose verbiage which is familiar to those users, but that's in addition to and not instead of, the current accepted `mdspan`-specific verbiage.